### PR TITLE
Don't leave an empty znode when replicated table is dropped

### LINKED ISSUE #61065: Don't leave behind an empty znode when dropping ReplicatedMergeTree
When dropping ReplicatedMergeTree in Replicated databse, we remove znode "/clickhouse/tables/{uuid}/{shard}" (default_replica_path), but keep the parent znode "/clickhouse/tables/{uuid}". If lots of tables are created and dropped, lots of empty znodes are left.

This will become more noticeable with refreshable materialized views, which create+drop a table on each refresh.

A possible solution is to make ReplicatedMergeTree delete parent znode too, if it became empty. The two znodes should probably be deleted in one transaction, otherwise there's a race with createAncestors(). If the double-delete transaction failed with ZNOTEMPTY, retry it with just one znode.

### PR DESCRIPTION
**Pull request by al13n321**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

Closes https://github.com/ClickHouse/ClickHouse/issues/61065

When dropping ReplicatedMergeTree, after removing znode at `zookeeper_path`, also remove its ancestors if they became empty, up through the path component with `{uuid}` macro. If there's no `{uuid}` macro, don't remove any ancestors.

E.g. if zookeeper path is `/foo/{uuid}/bar/baz`, table drop will remove `/foo/{uuid}/bar/baz` (if all replicas were dropped), then remove `/foo/{uuid}/bar` if it has no children left, then `/foo/{uuid}` if it has no children left.

Also extracted some of the zookeeper_path logic (expanding macros etc) into functions to be reused by refreshable materialized view in a later PR - refreshable MV coordination znode (https://github.com/ClickHouse/ClickHouse/pull/60669) needs to be managed in a similar way.

Most of the PR is refactoring, the new behavior is in `TableZnodeInfo::dropAncestorZnodesIfNeeded`.

### CODE REVIEW

**Comment by robot-ch-test-poll4** [comment]
*This is an automated comment for commit 9a3adc70bde322f67f8ef3ba5e21847c0732bbf9 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/69574/9a3adc70bde322f67f8ef3ba5e21847c0732bbf9/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/f62cb32c5c9f94e4d661b6a00c012564af5200f7/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

**Comment by al13n321** [comment]
* 02532_send_logs_level_test - flaky
* AST fuzzer (ubsan) - I couldn't figure out how to extract any useful information about the failure out of github actions

File: src/Storages/TableZnodeInfo.cpp [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+    String path_to_remove = path;
+    while (path_to_remove.size() > path_prefix_for_drop.size())
+    {
+        size_t i = path_to_remove.find_last_of('/');
+        chassert(i != String::npos && i >= path_prefix_for_drop.size());
+        path_to_remove = path_to_remove.substr(0, i);
+
+        Coordination::Error rc = zookeeper->tryRemove(path_to_remove);
+        if (rc != Coordination::Error::ZOK)
+            /// Znode not empty or already removed by someone else.
```
Comment (jkartseva) [inline src/Storages/TableZnodeInfo.cpp:130]: Wdyt about checking the return code and printing a debug message if it's neither `ZNOTEMPTY` nor `ZNONODE`.

File: src/Storages/StorageReplicatedMergeTree.h [hunk 2]
```diff
@@ -244,14 +243,15 @@ class StorageReplicatedMergeTree final : public MergeTreeData
...
                             LoggerPtr logger, MergeTreeSettingsPtr table_settings = nullptr, std::optional<bool> * has_metadata_out = nullptr);
 
-    bool dropReplica(const String & drop_zookeeper_path, const String & drop_replica, LoggerPtr logger);
+    bool dropReplica(const String & drop_replica, LoggerPtr logger);
 
     /// Removes table from ZooKeeper after the last replica was dropped
-    static bool removeTableNodesFromZooKeeper(zkutil::ZooKeeperPtr zookeeper, const String & zookeeper_path,
-                                              const zkutil::EphemeralNodeHolder::Ptr & metadata_drop_lock, LoggerPtr logger);
+    static bool removeTableNodesFromZooKeeper(
+        zkutil::ZooKeeperPtr zookeeper, const TableZnodeInfo & zookeeper_info2,
```
Comment (jkartseva) [inline src/Storages/StorageReplicatedMergeTree.h:253]: Why not name the parameter `zookeeper_info`? It's a static method, and the name won't conflict with the class field.

File: src/Storages/TableZnodeInfo.cpp [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+        ///  expansion passes is sketchy too (error-prone and more complex), and on balance this seems
+        ///  better.)
+        String uuid_str = toString(table_id.uuid);
+        size_t i = res.path.rfind(uuid_str);
+        if (i == String::npos)
+            /// Possible if the macro is in the "<auxiliary_zookeeper_name>:/" prefix, but we probably
+            /// don't want to allow that.
+            throw Exception(ErrorCodes::LOGICAL_ERROR, "Can't find uuid in zookeeper path after expanding {{uuid}} macro: {} (uuid {})", res.path, uuid_str);
+        i += uuid_str.size();
+        /// In case the path is "/foo/pika{uuid}chu/bar" (or "/foo/{uuid}{replica}/bar").
```
Comment (jkartseva) [inline src/Storages/TableZnodeInfo.cpp:105]: Should a `/foo/pikachu{uuid}/` prefix also be filtered out?

**Review by jkartseva — APPROVED** [review]
Looks good

File: src/Storages/TableZnodeInfo.cpp (hunk 1 above)
Comment (al13n321) [reply src/Storages/TableZnodeInfo.cpp:130]: `tryRemove` throws in that case:
```
    /// Doesn't throw in the following cases:
    /// * The node doesn't exist
    /// * Versions don't match
    /// * The node has children.
```
("Versions don't match" should never happen if version = -1, but client doesn't check that; maybe it should.)

File: src/Storages/TableZnodeInfo.cpp (hunk 3 above)
Comment (al13n321) [reply src/Storages/TableZnodeInfo.cpp:105]: `i` will end up at the end of the path component containing the uuid (right?), and that's what we want (right?). `{A}/{B}{uuid}{C}/{D}` -> `{A}/{B}{uuid}{C}`, for any (possibly empty) strings A, B, C, D (A and B not containing slashes).
 * `/foo/pikachu{uuid}/[...]` -> `/foo/pikachu{uuid}`
 * `/foo/pika{uuid}pika/[...]` -> `/foo/pika{uuid}pika`

**Pull request merged**